docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,129 @@
+  NodeCommands cmds ->
+    runNodeCmds cmds
+      & firstExceptT NodeCmdError
+  ByronCommand cmds ->
+    firstExceptT ByronClientError $ runByronClientCommand cmds
+  CompatibleCommands cmd ->
+    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
+      runAnyCompatibleCommand cmd
+...
+```
+- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.
+
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Proposed Solution
+
+Use the RIO monad which is based on the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).
+
+### Pros 
+I opt for the `RIO env a` monad instead of `ReaderT e m a` because: 
+- We get additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
+
+### Cons
+- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
+- Implicit error flow. Errors are thrown via `IO` exceptions.
+
+### Exception handling
+
+#### Exception type
+```haskell
+data CustomCliException where
+  CustomCliException
+    :: (Show error, Typeable error, Error error, HasCallStack)
+    => error -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+
+#### Top level handler 
+```haskell
+executeRio :: RIO () () -> IO ()
+executeRio r = do
+  runRIO () r
+    `catch` ( \e ->
+                case fromException e of
+                  Just custom@(CustomCliException{}) -> putStrLn $ displayException custom
+                  Nothing -> putStrLn $ displayException e
+            )
+```
+
+### Exception Handling Mechanism: Pros & Cons
+
+#### Pros
+
+1. **Unified Exception Type**
+```haskell
+data CustomCliException where
+  CustomCliException :: (Show e, Typeable e, Error e, HasCallStack) => e -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+- Simplifies Top-Level Handling: All errors are caught as `CustomCliException`.
```

Comment:
This ADR does not explain how this type should be used. You can't `fromException` other exceptions into `CustomCliExceptions`, so it will be useful only when explicitly thrown.

Reply:
>  You can't fromException other exceptions into CustomCliExceptions

I don't understand this objection. Can you clarify?


> Is this a temporary solution to bridge between ExceptT errors and exceptions, or a final design?

This is a _suggestion_ of a final design.

The purpose of `CustomCliException` is to represent explicitly thrown, structured errors that are meaningful to our application. For other exceptions (e.g., runtime errors or unexpected failures), we rely on the `Nothing` case in our handler.

Reply:
>I don't understand this objection. Can you clarify?

You've answered my question in the next paragraph of your comment. 👍🏻 

>This is a suggestion of a final design.

>The purpose of CustomCliException is to represent explicitly thrown, structured errors that are meaningful to our application. For other exceptions (e.g., runtime errors or unexpected failures), we rely on the Nothing case in our handler.

Ok. The thing is, that explanation should be in the ADR itself, so that it's clear for the next person reading it.

Reply:
It will be a strict rule when the ADR gets merged. I.e I want the team's buy in.

Reply:
>All errors are caught as CustomCliException.

This is nowhere demonstrated.

Reply:
![image](https://github.com/user-attachments/assets/d8dac19d-5653-4470-81f8-b1b56f130a3e)

Reply:
https://github.com/input-output-hk/cardano-node-wiki/pull/60#discussion_r1943239307